### PR TITLE
[#100906588] Implementing URL attachments as in the gengo-php clint library

### PR DESCRIPTION
**Pull request by marexandre**
Implementing `url_attachments` for handling URL based attachments for order and jobs comments.
Reflecting gengo-php clint library changes: https://github.com/gengo/gengo-php/pull/35

### CODE REVIEW

**Comment by hayate** [comment]
LGTM

File: gengo/gengo.py [hunk 1]
```diff
@@ -270,19 +272,19 @@ def get(self, **kwargs):
                             j['file_key'] = 'file_' + k
                             del j['file_path']
 
-            # If any attachments then modify base url to include
-            # private_key and file_data to include attachments as multipart
-            files = []
-            if 'attachments' in post_data:
+            # If any files then modify base url to include
+            # private_key and file_data to include files as multipart
+            tmpFiles = []
```
Comment (narusemotoki) [inline gengo/gengo.py]: I think snake_case is the best in this repository.

Comment (marexandre) [inline gengo/gengo.py]: Thanks! :bow:

File: gengo/gengo.py [hunk 2]
```diff
@@ -270,19 +272,19 @@ def get(self, **kwargs):
...
                             del j['file_path']
 
-            # If any attachments then modify base url to include
-            # private_key and file_data to include attachments as multipart
-            files = []
-            if 'attachments' in post_data:
+            # If any files then modify base url to include
+            # private_key and file_data to include files as multipart
+            tmp_files = []
+            if 'files' in post_data:
```
Comment (yugui) [inline gengo/gengo.py]: Could you move this `if` clause into the `try` clause?
The original implementation does not close other files if it fails to open one of the files.

**Comment by yugui** [comment]
LGTM

**Comment by marexandre** [comment]
Need to make changes related to `gengo-php` client changes

**Comment by marexandre** [comment]
Updates the code to handle `url-attachments` as we do in gengo-php client library. At the moment no actual file attachment is supported, only URL attachments. When we decide on the file attachments we would need to update the client library :pray:

**Comment by hayate** [comment]
LGTM

File: gengo/gengo.py [hunk 3]
```diff
@@ -399,6 +417,22 @@ def signAndRequestAPILatest(self, fn, base, query_params, post_data={},
...
 
+    def modifyURLAttachments(self, obj):
+        """
+        modifyURLAttachments handles url attachments for a given object
+
+        obj - job or comment object
+        """
+        if 'url_attachments' in obj:
+            if not isinstance(obj['url_attachments'], list):
+                raise GengoError("Job url attachment MUST be an array", 1)
```
Comment (narusemotoki) [inline gengo/gengo.py]: `array` -> `list` ?

File: gengo/gengo.py [hunk 4]
```diff
@@ -399,6 +417,22 @@ def signAndRequestAPILatest(self, fn, base, query_params, post_data={},
                               # SSL here ...
                               verify=False)
 
+    def modifyURLAttachments(self, obj):
+        """
+        modifyURLAttachments handles url attachments for a given object
```
Comment (yugui) [inline gengo/gengo.py]: describe what purpose it modifies for.
reflect its responsibility to the method name.

File: gengo/gengo.py [hunk 5]
```diff
@@ -399,6 +417,22 @@ def signAndRequestAPILatest(self, fn, base, query_params, post_data={},
...
+
+        obj - job or comment object
+        """
+        if 'url_attachments' in obj:
+            if not isinstance(obj['url_attachments'], list):
+                raise GengoError("Job url attachment MUST be an list", 1)
+
+            obj['attachments'] = []
```
Comment (yugui) [inline gengo/gengo.py]: `obj['attachments'] = obj['url_attachments'][:]`?

or even `obj['attachments'] = obj['url_attachments']`?

File: gengo/gengo.py [hunk 6]
```diff
@@ -206,14 +206,17 @@ def get(self, **kwargs):
                     post_data['jobs']['as_group'] = jobs_dict.pop('as_group')
                 if 'comment' in jobs_dict:
                     post_data['jobs']['comment'] = jobs_dict.pop('comment')
+                if 'url_attachments' in jobs_dict:
```
Comment (yugui) [inline gengo/gengo.py]: it looks like more than refactoring.  What is the purpose of this PR?

Comment (marexandre) [inline gengo/gengo.py]: Update the PR's title and description

**Comment by hayate** [comment]
LGTM

**Comment by yugui** [comment]
LGTM

**Comment by hayate** [comment]
@marexandre cold you please update the change log with the new functionality, it should be a brief description.

**Comment by marexandre** [comment]
@hayate As I merged this PR, here is the new PR https://github.com/gengo/gengo-python/pull/60 :bow:

**Pull request merged**